File: python/树/1339. 分裂二叉树的最大乘积.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    res = set()
    def maxProduct(self, root: TreeNode) -> int:
        num = self.get_sum(root)
        ans = 0
        for i in self.res:
            ans = max(ans,i*(num-i))
        ans %= 10**9+7
        return ans

    def get_sum(self,root):
        if not root: return 0
        val = root.val + self.get_sum(root.left) + self.get_sum(root.right)
        self.res.add(val)
        return val

File: go/tree/1339. 分裂二叉树的最大乘积.go
package tree

import "math"

var res map[int]struct{}

func maxProduct(root *TreeNode) int {
	res = make(map[int]struct{})
	num := getSum(root)
	ans := 0
	for k, _ := range res {
		if ans > k*(num-k) {
			continue
		} else {
			ans = k * (num - k)
		}
	}
	return ans % (int(math.Pow(10, 9)) + 7)
}

func getSum(root *TreeNode) int {
	if root == nil {
		return 0
	}
	val := root.Val + getSum(root.Left) + getSum(root.Right)
	res[val] = struct{}{}
	return val
}
